### ISSUE TITLE
out-structs containing objects and refcounting

### ISSUE THREAD

**Issue by Kangz** [webgpu-native/webgpu-headers#326]
We should define what happens to objects contained in out-structs that will be overwritten. I had the following code and it was leaking quite unexpectedly:

```c
WGPUSurfaceTexture s;

wgpuSurfaceGetCurrentTexture(&s);
wgpuSurfacePresent();
wgpuSurfaceGetCurrentTexture(&s);

wgpuSurfaceTextureFreeMembers(&s);
```

The issue is that the second `wgpuSurfaceGetCurrentTexture` call overwrites `s.texture` without decrementing the refcount, which means that `wgpuSurfaceTextureFreeMembers` doesn't know to unref it. (or maybe it is a C++ism of Dawn that we need to fix, but in all cases the behavior is surprising).

In all cases I'd suggest that if an object is passed in an out-struct, it must be either null or valid, and if valid it will be released once.

**Comment by kainino0x** [comment]
`wgpuSurfaceTextureFreeMembers` doesn't exist (I got confused about this at some point) because there are no arrays of data to free.

Generally speaking, just because there's a pointer value for a WGPUTexture in WGPUSurfaceTexture doesn't mean that's a valid reference. It could have been Released already and the app didn't zero out the pointer because they knew they were about to overwrite it:

```c
wgpuTextureRelease(s.texture);
wgpuSurfaceGetCurrentTexture(&s); // double free
```

I think it's the reality of the C API that we don't know what the intent was so we should just overwrite the pointer without releasing it. It's unfortunate that the overwrite is hard to see (vs. if we just did `texture = wgpuSurfaceGetCurrentTexture();` which is more obviously a leak) but I think it's fine.

In our C++ API, we _do_ know whether the ref is still there because the RAII wrapper already needs to know. If we implemented the C++ API in C++ that would happen automatically, because we'd write `s->texture = nextTexture;` and that would release the existing ref. I think it's just because of our C/C++ reinterpret_casting that we don't currently do that, so we would need to do `s->texture = {};` before calling from the C++ wrapper into the C API.
